### PR TITLE
Make bool_validator strict

### PR DESCRIPTION
**Pull request by dmontagu**
## Change Summary

Make `bool_validator` a little more strict, in line with https://github.com/samuelcolvin/pydantic/issues/579

In particular, allow `True`, `False`, `1`, `0`, all yaml true/false strings, the strings `'t'`/`'T'`/`'f'`/`'F'` (following DRF), and the bytes versions of those strings. Anything else raises a `ValidationError`.

Also adds `RelaxedBool` to retain current functionality if desired.

## Related issue number

#579

------

Reviewing the discussion in #579, it wasn't clear to me whether the goal was to modify `StrictBool` to allow these strings, or to change `bool_validator`. My current interpretation is that the goal was to change `bool_validator` to bring it in line with more expected/useful behavior in a non-python context. If that's wrong, I'd be happy to refactor so that this logic applies to `StrictBool` instead.

This would be a breaking change for a very common type (if modifying `bool_validator` anyway), so I wouldn't necessarily expect it to be merged prior to a 1.0 release, but figured I'd start the process.

------

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

### CODE REVIEW

File: pydantic/validators.py [hunk 1]
```diff
@@ -80,17 +80,20 @@ def bytes_validator(v: Any) -> bytes:
         raise errors.BytesError()
 
 
-BOOL_STRINGS = {'1', 'TRUE', 'ON', 'YES'}
+BOOL_TRUE = {True, 1, 'y', 'Y', 'yes', 'Yes', 'YES', 'true', 'True', 'TRUE', 'on', 'On', 'ON', '1', 't', 'T'}
+BOOL_FALSE = {False, 0, 'n', 'N', 'no', 'No', 'NO', 'false', 'False', 'FALSE', 'off', 'Off', 'OFF', '0', 'f', 'F'}
 
 
 def bool_validator(v: Any) -> bool:
     if isinstance(v, bool):
```
Comment (dmontagu) [inline pydantic/validators.py]: I kept this logic fork (even though redundant) so you can retain maximum performance by using actual bools.

File: tests/test_types.py [hunk 2]
```diff
@@ -315,18 +315,47 @@ class Config:
...
+        ('bool_check', 'F', False),
+        ('bool_check', b'FALSE', False),
+        ('bool_check', None, ValidationError),
+        ('bool_check', '', ValidationError),
+        ('bool_check', [], ValidationError),
+        ('bool_check', {}, ValidationError),
+        ('bool_check', [True], ValidationError),
+        ('bool_check', b'2', ValidationError),
+        ('bool_check', '2', ValidationError),
+        ('bool_check', 2, ValidationError),
```
Comment (dmontagu) [inline tests/test_types.py]: I replaced the test cases with all allowed values excluding bytes versions, along with one bytes version each for True and False. I also added what I figured would be the most common error cases.

File: pydantic/validators.py [hunk 3]
```diff
@@ -80,17 +80,20 @@ def bytes_validator(v: Any) -> bytes:
         raise errors.BytesError()
 
 
-BOOL_STRINGS = {'1', 'TRUE', 'ON', 'YES'}
+BOOL_TRUE = {True, 1, 'y', 'Y', 'yes', 'Yes', 'YES', 'true', 'True', 'TRUE', 'on', 'On', 'ON', '1', 't', 'T'}
```
Comment (samuelcolvin) [inline pydantic/validators.py]: wouldn't it be simpler if we just did `v.lower() in BOOL_TRUE` and removed the case duplicates?

I get that this would mean `'trUE' -> True` but does anyone care?

Comment (cazgp) [reply pydantic/validators.py]: Yup I agree

File: pydantic/validators.py [hunk 4]
```diff
@@ -80,17 +80,20 @@ def bytes_validator(v: Any) -> bytes:
...
     if isinstance(v, bool):
         return v
     if isinstance(v, bytes):
         v = v.decode()
-    if isinstance(v, str):
-        return v.upper() in BOOL_STRINGS
-    return bool(v)
+    if v in BOOL_TRUE:
+        return True
+    elif v in BOOL_FALSE:
```
Comment (cazgp) [inline pydantic/validators.py]: Unnecessary `elif` and breaks consistency with the rest of the function

File: tests/test_types.py (hunk 2 above)
Comment (cazgp) [reply tests/test_types.py]: This is missing the case for a dict of things, and an array of things.

```
bool({}) == False
bool({"test": 1}) == True
bool([]) == False
bool([1]) == True
```

**Comment by cazgp** [comment]
Thanks for this!

File: tests/test_types.py (hunk 2 above)
Comment (samuelcolvin) [reply tests/test_types.py]: I guess for completeness and "tests as a definition of behaviour" we should also include a test for an instance of a class with an explicit `__bool__` method.

File: pydantic/validators.py (hunk 3 above)
Comment (dmontagu) [reply pydantic/validators.py]: I'll change it

File: tests/test_types.py (hunk 2 above)
Comment (dmontagu) [reply tests/test_types.py]: @samuelcolvin if it has an explicit `__bool__` method, should it be used? Or should it fail validation? I'd expect it to be used in this non-strict case, but just want to confirm first.

Comment (samuelcolvin) [reply tests/test_types.py]: well that's debatable, I guess personally I think it should be be accepted by I know @cazgp would like it to be stricter.

I guess it doesn't matter that much since nothing coming from "dumb deserialization" eg. JSON will have such a method.

Comment (dmontagu) [reply tests/test_types.py]: @samuelcolvin 
```python
Python 3.7.3
>>> hasattr(2, "__bool__")
True
>>> hasattr(2.5, "__bool__")
True
```
given this, how would you propose to handle the `__bool__` attribute? I could add checks for int / float and handle these cases appropriately, but this has me worried about unexpected edge cases and/or needing to handle a long list of them.

 @cazgp Any thoughts? (I'm assuming you are in favor of dropping `__bool__` support, but figured I'd ask)

Comment (samuelcolvin) [reply tests/test_types.py]: In that case I think it would be easiest to not include `__bool__` support.

File: tests/test_types.py [hunk 5]
```diff
@@ -316,22 +316,75 @@ class Config:
...
+#
+# class BoolCastableModel(BaseModel):
+#     x: int
+#
+#     def __bool__(self) -> bool:
+#         return self.x > 0
+#
+# class NonBoolCastableModel(BaseModel):
+#     x: int
+
```
Comment (dmontagu) [inline tests/test_types.py]: I will replace these commented classes with appropriate alternatives (and alter/remove references in the non-test code) once we have a consensus on `__bool__` handling.

File: tests/test_types.py (hunk 2 above)
Comment (cazgp) [reply tests/test_types.py]: Yeah, `__bool__` seems way too open and if it undermines all of the above we should definitely not take it into account!

Comment (samuelcolvin) [reply tests/test_types.py]: Fine by me, but I propose a sweepstake on how long it will take for someone to request `RelaxedBool` that behaves like python's `bool`.

My guess - within a week of this being released.

Comment (dmontagu) [reply tests/test_types.py]: @samuelcolvin I got in front of that and just added RelaxedBool. Let me know if you want me to remove it

**Comment by dmontagu** [comment]
@samuelcolvin I've updated the docs. However, I still have two questions:
1) should I update the history? I'm not sure if the goal is to include this in an upcoming 0.x release or wait for 1.0
2) I only made note of the change in the section of the docs about `StrictBool` and `RelaxedBool`. I think this is okay, but given this is a breaking change to the behavior of `bool`, should a note be placed somewhere else as well?

File: docs/index.rst [hunk 6]
```diff
@@ -567,13 +567,25 @@ Exotic Types
...
+A standard ``bool`` field will raise a ``ValidationError`` if the value is not one of the following:
+
+* A valid boolean (i.e., ``True`` or ``False``),
+* The integers ``0`` or ``1``,
+* a ``str`` which when lowercased is in ``{'off', 'f', 'false', 'n', 'no', '1', 'on', 't', 'true', 'y', 'yes'}``
+* a ``bytes`` which is valid (per the previous rule) when decoded to ``str``
+
+If you want more permissive behavior, ``RelaxedBool`` can be used to cast values to bool using standard python logic.
+For example, ``None`` would be cast to False. Note that this behaves differently than the standard ``bool`` field
+for some values: for example, a ``RelaxedBool`` field would cast the string "false" to True.
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
for some values: for example, a ``RelaxedBool`` field would cast the string ``"false"`` to ``True``.
```

File: docs/index.rst [hunk 7]
```diff
@@ -567,13 +567,25 @@ Exotic Types
...
+
+For stricter behavior, ``StrictBool`` can be used to require specifically ``True`` or ``False``;
 nothing else is permitted.
 
 
+
```
Comment (samuelcolvin) [inline docs/index.rst]:
```suggestion
```

File: pydantic/validators.py [hunk 8]
```diff
@@ -99,8 +100,13 @@ def bool_validator(v: Any) -> bool:
     if isinstance(v, bytes):
         v = v.decode()
     if isinstance(v, str):
-        return v.upper() in BOOL_STRINGS
-    return bool(v)
+        v = v.lower()
+    with change_exception(errors.BoolError, TypeError):
```
Comment (samuelcolvin) [inline pydantic/validators.py]: can we have a test for this failure.

File: docs/index.rst [hunk 9]
```diff
@@ -567,13 +567,25 @@ Exotic Types
 (This script is complete, it should run "as is")
 
 
-StrictBool
-~~~~~~~~~~
+RelaxedBool and StrictBool
```
Comment (samuelcolvin) [inline docs/index.rst]: Let's:
* change this to "RelaxedBool, StrictBool and bool" or just "Booleans"
* move the examples into a separate code example here
* add a warning explaining how bool behaviour changed in the new version.

File: pydantic/validators.py (hunk 8 above)
Comment (dmontagu) [reply pydantic/validators.py]: Done (the test is called `test_bool_unhashable_fails`); let me know if I've misinterpreted your intention.

**Comment by samuelcolvin** [comment]
please can you rebase and move the history change to `changes/` as per #719.

File: pydantic/types.py [hunk 10]
```diff
@@ -243,6 +244,21 @@ def validate(cls, value: str) -> str:
         return validate_email(value)[1]
 
 
+class RelaxedBool(int):
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/types.py]: I think the only reason to use it would be performance; right now if you want speed when you *know* you are getting the right type, it's awkward. But I agree it would be better if the name reflected this.

I'd be okay either dropping or giving a different name.

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/types.py]: okay, let's keep it but say that in the docs.

File: docs/index.rst [hunk 11]
```diff
@@ -580,12 +580,35 @@ Exotic Types
...
 
-StrictBool
-~~~~~~~~~~
+Booleans
+~~~~~~~~
+
+.. warning::
+
+    The logic for parsing ``bool`` fields has changed as of version v0.31.
+    Prior to v0.31, ``bool`` fields would be parsed by simply mapping ``value`` to ``bool(value)``.
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply docs/index.rst]: Yes, I realized that recently 🤦‍♂️

File: pydantic/types.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/types.py]: @samuelcolvin I did some minor investigation of the timing and I think the speed difference between the two bool validators is negligible, possibly with the standard one actually being faster than the relaxed one.

I think probably the only practical benefit is that you'd get to use a custom `__bool__` implementation if you wanted to use an arbitrary type in a field.

So I think the only thing to say in the docs is that it follows built-in python logic for boolean casting (what they currently say). Let me know if you still disagree, or if this changes your mind back to wanting to remove `RelaxedBool`.

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/types.py]: I would say let's remove it then. Nothing to stop people implementing it themselves if they really want it. But we we think there's basically no context in which you should use it, it shouldn't be added.

Sorry to flip flop.

File: pydantic/validators.py [hunk 12]
```diff
@@ -100,8 +101,13 @@ def bool_validator(v: Any) -> bool:
     if isinstance(v, bytes):
         v = v.decode()
```
[the remark this answers is not in the document]
Comment (dmontagu) [reply pydantic/validators.py]: `UnicodeDecodeError` is a subclass of `ValueError`, so it gets handled properly already. Do you still want me to explicitly handle it here?

[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/validators.py]: No, I guess test is enough.
